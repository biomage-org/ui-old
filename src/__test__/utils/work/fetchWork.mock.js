const mockCacheKeyMappings = {
  A: 'fd3161a878f67ebf54018720cffd6a66', // pragma: allowlist secret
  B: '10250a11679234110a1c260d6fd81d3c', // pragma: allowlist secret
  C: '76bf160685c4c80c67abd9a701da23e6', // pragma: allowlist secret
  D: '21671038b9ac73c1f08a94c8213cb872', // pragma: allowlist secret
  E: '33faa711a94a2028b5bae1778126aec0', // pragma: allowlist secret
};

const mockGenesListData = ['A', 'B', 'C', 'D'];

const mockGeneExpressionData = {
  A: {
    rawExpression: {
      min: 0,
      max: 6.8,
      mean: 1.68,
      stdev: 2.597331964,
      expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
    },
    truncatedExpression: {
      min: 0,
      max: 6.8,
      mean: 1.68,
      stdev: 2.597331964,
      expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
    },
  },
  B: {
    rawExpression: {
      min: 0,
      max: 6.8,
      mean: 1.702857143,
      stdev: 2.551115536,
      expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
    },
    truncatedExpression: {
      min: 0,
      max: 6.8,
      mean: 1.702857143,
      stdev: 2.551115536,
      expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
    },
  },
  C: {
    rawExpression: {
      min: 0,
      max: 3.4,
      mean: 1.68,
      stdev: 2.141525936,
      expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
    },
    truncatedExpression: {
      min: 0,
      max: 3.4,
      mean: 1.68,
      stdev: 2.141525936,
      expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
    },
  },
  D: {
    rawExpression: {
      min: 0,
      max: 6.8,
      mean: 1.68,
      stdev: 2.597331964,
      expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
    },
    truncatedExpression: {
      min: 0,
      max: 6.8,
      mean: 1.68,
      stdev: 2.597331964,
      expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
    },
  },
  E: { hello: 'world' },
};

const mockGeneExpressionDataZScore = {
  D: {
    zScore: [
      -0.6468175894669735,
      -0.6468175894669735,
      -0.6468175894669735,
      -0.04620125639049807,
      -0.6468175894669735,
      1.971253605994586,
      0.6622180082638063,
    ],
  },
};

const mockCacheGet = jest.fn((x) => {
  const mockCacheContents = {
    fd3161a878f67ebf54018720cffd6a66: 'A', // pragma: allowlist secret
    '10250a11679234110a1c260d6fd81d3c': 'B', // pragma: allowlist secret
    f5c957411a28de68f35e1f5c8a29da7e: 'C', // pragma: allowlist secret
    '33faa711a94a2028b5bae1778126aec0': 'E', // pragma: allowlist secret
  };

  if (x in mockCacheContents) {
    return mockGeneExpressionData[mockCacheContents[x]];
  }

  return null;
});
const mockCacheSet = jest.fn();
const mockCacheRemove = jest.fn();

const mockDispatchWorkRequest = jest.fn(() => true);

const mockSeekFromS3 = jest.fn();

const mockCacheModule = {
  get: jest.fn((x) => mockCacheGet(x)),
  set: jest.fn((key, val) => mockCacheSet(key, val)),
  _remove: jest.fn((key) => mockCacheRemove(key)),
};

const mockSeekWorkResponseModule = {
  __esModule: true,
  seekFromS3: mockSeekFromS3,
  dispatchWorkRequest: mockDispatchWorkRequest,
};

const mockQcPipelineStartDate = '2021-01-01T01:01:01.000Z';

const mockReduxState = (experimentId, environment = 'testing') => () => ({
  networkResources: {
    environment,
  },
  backendStatus: {
    [experimentId]: {
      status: {
        pipeline: {
          status: 'SUCCEEDED',
          startDate: mockQcPipelineStartDate,
        },
      },
    },
  },
});

export {
  mockGeneExpressionData, mockGeneExpressionDataZScore,
  mockGenesListData,
  mockCacheKeyMappings,
  mockCacheGet, mockCacheSet,
  mockDispatchWorkRequest, mockSeekFromS3,
  mockReduxState, mockQcPipelineStartDate,
  mockCacheModule, mockSeekWorkResponseModule,
};
