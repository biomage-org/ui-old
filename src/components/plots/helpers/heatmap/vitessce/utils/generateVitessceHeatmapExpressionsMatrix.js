import _ from 'lodash';
import { convertRange } from 'utils/plotUtils';

const scaledTo255 = (rowOfExpressions, min, max) => (
  rowOfExpressions.map((value) => convertRange(value, [min, max], [0, 255]))
);

const generateVitessceHeatmapExpressionsMatrix = (cellOrder, geneOrder, expressionMatrix) => {
  const geneExpressionsDataMatrix = [];

  geneOrder.forEach((gene) => {
    if (!expressionMatrix.geneIsLoaded(gene)) return;
    const truncatedExpression = expressionMatrix.getTruncatedExpression(gene, cellOrder);

    const { truncatedMin, truncatedMax } = expressionMatrix.getStats(gene);

    const scaledGeneExpressions = scaledTo255(truncatedExpression, truncatedMin, truncatedMax);

    geneExpressionsDataMatrix.push(scaledGeneExpressions);
  });

  return _.flatten(_.unzip(geneExpressionsDataMatrix));
};

export default generateVitessceHeatmapExpressionsMatrix;
