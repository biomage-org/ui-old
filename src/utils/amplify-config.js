import Environment, { ssrGetCurrentEnvironment } from './environment';
import getAccountId from './getAccountId';

const configure = (userPoolId, identityPoolId, userPoolClientDetails) => {
  const currentEnvironment = ssrGetCurrentEnvironment();
  const accountId = getAccountId(currentEnvironment);
  const bucketName = `biomage-originals-${currentEnvironment}-${accountId}`;

  const storageConfig = {
    Storage: {
      AWSS3: {
        bucket: bucketName,
        region: 'eu-west-1',
        dangerouslyConnectToHttpEndpointForTesting: currentEnvironment === Environment.DEVELOPMENT,
        identityId: identityPoolId,
        customPrefix: {
          public: '',
        },
      },
    },
  };
  const redirectProtocol = (process.env.NODE_ENV === 'development') ? 'http:' : 'https:';
  const usingProtocol = (url) => url.startsWith(redirectProtocol);
  const signInRedirect = userPoolClientDetails.CallbackURLs.filter(usingProtocol)[0];
  const signOutRedirect = userPoolClientDetails.LogoutURLs.filter(usingProtocol)[0];

  const authConfig = {
    Auth: {
      region: 'eu-west-1',
      identityPoolId,
      userPoolId,
      userPoolWebClientId: userPoolClientDetails.ClientId,

      mandatorySignIn: false,
      authenticationFlowType: 'USER_SRP_AUTH',

      oauth: {
        domain: userPoolClientDetails.Domain,
        scope: userPoolClientDetails.AllowedOAuthScopes,
        redirectSignIn: signInRedirect,
        redirectSignOut: signOutRedirect,
        responseType: userPoolClientDetails.AllowedOAuthFlows[0],
      },
    },
  };

  return (
    { ...authConfig, ...storageConfig }
  );
};

export default configure;
