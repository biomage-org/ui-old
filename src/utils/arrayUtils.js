import _ from 'lodash';

const isSubset = (
  (subsetArray, containingArray) => _.difference(subsetArray, containingArray).length === 0);

// Like reverse but it doesn't mutate underlying array
const reversed = (array) => [...array].reverse();

const removed = (item, array) => array.filter((currExpId) => currExpId !== item);

export { isSubset, reversed, removed };
