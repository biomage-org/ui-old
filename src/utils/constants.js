const api = {
  V1: 'v1',
  V2: 'v2',
};

const modules = {
  DATA_MANAGEMENT: 'DATA_MANAGEMENT',
  DATA_PROCESSING: 'DATA_PROCESSING',
  DATA_EXPLORATION: 'DATA_EXPLORATION',
  PLOTS_AND_TABLES: 'PLOTS_AND_TABLES',
  SETTINGS: 'SETTINGS',
  DEFAULT: '',
};

const sampleTech = {
  '10X': '10x',
  RHAPSODY: 'rhapsody',
};

const plotTypes = {
  CONTINUOUS_EMBEDDING: 'ContinuousEmbedding',
  CATEGORICAL_EMBEDDING: 'CategoricalEmbedding',
  HEATMAP: 'Heatmap',
  MARKER_HEATMAP: 'MarkerHeatmap',
  VOLCANO_PLOT: 'VolcanoPlot',
  FREQUENCY_PLOT: 'FrequencyPlot',
  VIOLIN_PLOT: 'violin',
  DOT_PLOT: 'DotPlot',
  TRAJECTORY_ANALYSIS: 'TrajectoryAnalysis',
  NORMALIZED_EXPRESSION_MATRIX: 'NormalizedExpressionMatrix',
  IMG_PLOT: 'ImgPlot',
};

const plotNames = {
  CONTINUOUS_EMBEDDING: 'Continuous Embedding',
  CATEGORICAL_EMBEDDING: 'Categorical Embedding',
  HEATMAP: 'Custom Heatmap',
  MARKER_HEATMAP: 'Marker Heatmap',
  VOLCANO_PLOT: 'Volcano Plot',
  FREQUENCY_PLOT: 'Frequency Plot',
  VIOLIN_PLOT: 'Violin Plot',
  DOT_PLOT: 'Dot Plot',
  TRAJECTORY_ANALYSIS: 'Trajectory Analysis',
  NORMALIZED_EXPRESSION_MATRIX: 'Normalized Expression Matrix',
  IMG_PLOT: 'Image plot'
};

const layout = {
  PANEL_HEADING_HEIGHT: 30,
  PANEL_PADDING: 10,
};

export {
  api,
  modules,
  sampleTech,
  plotTypes,
  plotNames,
  layout,
};
