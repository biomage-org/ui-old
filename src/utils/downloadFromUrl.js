import { isNil } from 'lodash';

const downloadFromUrl = (url, filename = null) => {
  const link = document.createElement('a');
  link.style.display = 'none';
  link.href = url;

  if (!isNil(filename)) link.download = filename;

  document.body.appendChild(link);
  link.click();

  setTimeout(() => {
    URL.revokeObjectURL(link.href);
    link.parentNode.removeChild(link);
  }, 0);
};

export default downloadFromUrl;
