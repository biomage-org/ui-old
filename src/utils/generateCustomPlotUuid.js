const generateDataProcessingPlotUuid = (sampleId, filterName, plotIdx) => {
  if (sampleId) {
    return `${sampleId}-${filterName}-${plotIdx}`;
  }
  return `${filterName}-${plotIdx}`;
};

const generateMultiViewGridPlotUuid = (plotUuid, plotIdx) => (
  `${plotUuid}-${plotIdx}`
);

export { generateDataProcessingPlotUuid, generateMultiViewGridPlotUuid };
