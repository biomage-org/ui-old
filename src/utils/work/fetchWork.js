/* eslint-disable no-underscore-dangle */
import { MD5 } from 'object-hash';

import config from 'config';

import { Environment, isBrowser } from 'utils/deploymentInfo';
import { getBackendStatus } from 'redux/selectors';

import cache from 'utils/cache';
import { dispatchWorkRequest, seekFromS3 } from 'utils/work/seekWorkResponse';

const createObjectHash = (object) => MD5(object);

// Disable unique keys to reallow reuse of work results in development
const DISABLE_UNIQUE_KEYS = [
  'GetEmbedding',
];

const generateETag = (
  experimentId,
  body,
  extras,
  qcPipelineStartDate,
  environment,
) => {
  // If caching is disabled, we add an additional randomized key to the hash so we never reuse
  // past results.
  let cacheUniquenessKey = null;

  if (
    environment !== Environment.PRODUCTION
    && localStorage.getItem('disableCache') === 'true'
    && !DISABLE_UNIQUE_KEYS.includes(body.name)
  ) {
    cacheUniquenessKey = Math.random();
  }

  let ETagBody = {
    experimentId,
    body,
    qcPipelineStartDate,
    extras,
    cacheUniquenessKey,
    workerVersion: config.workerVersion,
  };

  // They `body` key to create ETAg for gene expression is different
  // from the others, causing the generated ETag to be different
  if (body.name === 'GeneExpression') {
    ETagBody = {
      experimentId,
      missingGenesBody: body,
      qcPipelineStartDate,
      extras,
      cacheUniquenessKey,
      workerVersion: config.workerVersion,
    };
  }

  return createObjectHash(ETagBody);
};

// const decomposeBody = async (body, experimentId) => {
//   const { genes: requestedGenes } = body;
//   const missingDataKeys = {};
//   const cachedData = {};

//   await Promise.all(requestedGenes.map(async (g) => {
//     const newBody = {
//       ...body,
//       genes: g,
//     };

//     const key = createObjectHash({ experimentId, newBody });
//     const data = await cache.get(key);

//     if (data) {
//       cachedData[g] = data;
//     } else {
//       missingDataKeys[g] = key;
//     }
//   }));

//   return { missingDataKeys, cachedData };
// };

// Temporarily using gene expression without local cache
const fetchGeneExpressionWorkWithoutLocalCache = async (
  experimentId,
  timeout,
  body,
  backendStatus,
  environment,
  broadcast,
  extras,
) => {
  // If new genes are needed, construct payload, try S3 for results,
  // and send out to worker if there's a miss.
  const { pipeline: { startDate: qcPipelineStartDate } } = backendStatus;

  const ETag = generateETag(
    experimentId,
    body,
    extras,
    qcPipelineStartDate,
    environment,
  );

  // Then, we may be able to find this in S3.
  const response = await seekFromS3(ETag, experimentId);

  if (response) return response;

  // If there is no response in S3, dispatch workRequest via the worker
  try {
    await dispatchWorkRequest(
      experimentId,
      body,
      timeout,
      ETag,
      {
        ETagPipelineRun: qcPipelineStartDate,
        broadcast,
        ...extras,
      },
    );
  } catch (error) {
    console.error('Error dispatching work request: ', error);
    throw error;
  }

  return await seekFromS3(ETag, experimentId);
};

// const fetchGeneExpressionWork = async (
//   experimentId,
//   timeout,
//   body,
//   backendStatus,
//   environment,
//   broadcast,
//   extras,
// ) => {
//   // Get only genes that are not already found in local storage.
//   const { missingDataKeys, cachedData } = await decomposeBody(body, experimentId);

//   const missingGenes = Object.keys(missingDataKeys);

//   if (missingGenes.length === 0) {
//     return cachedData;
//   }

//   // If new genes are needed, construct payload, try S3 for results,
//   // and send out to worker if there's a miss.
//   const { pipeline: { startDate: qcPipelineStartDate } } = backendStatus;

//   const missingGenesBody = { ...body, genes: missingGenes };

//   const ETag = generateETag(
//     experimentId,
//     missingGenesBody,
//     extras,
//     qcPipelineStartDate,
//     environment,
//   );

//   // Then, we may be able to find this in S3.
//   let response = await seekFromS3(ETag, experimentId);

//   // If there is no response in S3, dispatch workRequest via the worker
//   if (!response) {
//     try {
//       await dispatchWorkRequest(
//         experimentId,
//         missingGenesBody,
//         timeout,
//         ETag,
//         {
//           ETagPipelineRun: qcPipelineStartDate,
//           broadcast,
//           ...extras,
//         },
//       );
//     } catch (error) {
//       console.error('Error dispatching work request: ', error);
//       throw error;
//     }
//   }

//   response = await seekFromS3(ETag, experimentId);

//   Object.keys(missingDataKeys).forEach(async (gene) => {
//     await cache.set(missingDataKeys[gene], response[gene]);
//   });

//   return response;
// };

const fetchWork = async (
  experimentId,
  body,
  getState,
  optionals = {},
) => {
  const {
    extras = undefined, timeout = 180, broadcast = false, fetchS3Data = true,
  } = optionals;
  const backendStatus = getBackendStatus(experimentId)(getState()).status;

  const { environment } = getState().networkResources;
  if (!isBrowser) {
    throw new Error('Disabling network interaction on server');
  }

  if (environment === Environment.DEVELOPMENT && !localStorage.getItem('disableCache')) {
    localStorage.setItem('disableCache', 'true');
  }

  const { pipeline: { startDate: qcPipelineStartDate } } = backendStatus;
  if (body.name === 'GeneExpression') {
    return fetchGeneExpressionWorkWithoutLocalCache(
      experimentId, timeout, body, backendStatus, environment, broadcast, extras,
    );
  }

  const ETag = generateETag(
    experimentId,
    body,
    extras,
    qcPipelineStartDate,
    environment,
  );

  // First, let's try to fetch this information from the local cache.
  const data = await cache.get(ETag);

  if (data) {
    return data;
  }

  console.log('BEFORE SEEK');
  // Then, we may be able to find this in S3.
  let response = await seekFromS3(ETag, experimentId, fetchS3Data);

  if (response) return response;

  // If there is no response in S3, dispatch workRequest via the worker
  try {
    await dispatchWorkRequest(
      experimentId,
      body,
      timeout,
      ETag,
      {
        PipelineRunETag: qcPipelineStartDate,
        broadcast,
        ...extras,
      },
    );

    response = await seekFromS3(ETag, experimentId, fetchS3Data);
  } catch (error) {
    console.error('Error dispatching work request', error);
    throw error;
  }

  // If a work response is in s3, it is cacheable
  // (the cacheable or not option is managed in the worker)
  await cache.set(ETag, response);
  return response;
};

export {
  fetchWork,
  generateETag,
};
