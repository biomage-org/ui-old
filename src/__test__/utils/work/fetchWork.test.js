/* eslint-disable global-require */
import { fetchWork } from 'utils/work/fetchWork';
import { Environment } from 'utils/deploymentInfo';

const {
  mockGeneExpressionData,
  mockGeneExpressionDataZScore,
  mockGenesListData,
  mockCacheKeyMappings,
  mockCacheGet,
  mockCacheSet,
  mockDispatchWorkRequest,
  mockSeekFromS3,
  mockReduxState,
} = require('__test__/utils/work/fetchWork.mock');

jest.mock('utils/cache', () => require('__test__/utils/work/fetchWork.mock').mockCacheModule);
jest.mock('utils/work/seekWorkResponse', () => require('__test__/utils/work/fetchWork.mock').mockSeekWorkResponseModule);

const experimentId = '1234';
const NON_GENE_EXPRESSION_ETAG = '013c3026bb7156d222ccd18919745195'; // pragma: allowlist secret
const GENE_EXPRESSION_ETAG = '34c05c9d07fd24ce0c22d2bec7fd7437'; // pragma: allowlist secret
const timeout = 10;

const nonGeneExpressionWorkRequest = {
  name: 'ListGenes',
};

const geneExpressionWorkRequest = {
  name: 'GeneExpression',
  genes: ['A', 'B', 'C', 'D'],
};

describe('fetchWork', () => {
  beforeEach(async () => {
    Storage.prototype.setItem = jest.fn();

    jest.clearAllMocks();

    mockSeekFromS3
      .mockReset()
      .mockImplementationOnce(() => null)
      .mockImplementationOnce(() => mockGenesListData);
  });

  it('runs correctly for gene expression work request', async () => {
    mockSeekFromS3
      .mockReset()
      .mockImplementationOnce(() => null)
      .mockImplementation(() => ({ D: mockGeneExpressionData.D }));

    const res = await fetchWork(
      experimentId,
      geneExpressionWorkRequest,
      mockReduxState(experimentId),
      { timeout },
    );

    expect(mockDispatchWorkRequest).toHaveBeenCalledWith(
      experimentId,
      { name: 'GeneExpression', genes: ['D'] },
      timeout,
      GENE_EXPRESSION_ETAG,
      expect.anything(),
    );

    // The expected response should contain the ZScore
    const expectedResponse = {
      D: {
        ...mockGeneExpressionData.D,
        ...mockGeneExpressionDataZScore.D,
      },
    };

    expect(mockCacheGet).toHaveBeenCalledTimes(4);
    expect(mockCacheSet).toHaveBeenCalledTimes(1);
    expect(mockCacheSet).toHaveBeenCalledWith(mockCacheKeyMappings.D, expectedResponse.D);
    expect(res).toEqual(expectedResponse);
    expect(mockSeekFromS3).toHaveBeenCalledTimes(2);
  });

  it('runs correctly for non gene expression work request', async () => {
    const res = await fetchWork(
      experimentId,
      nonGeneExpressionWorkRequest,
      mockReduxState(experimentId),
      { timeout: 10 },
    );

    expect(mockDispatchWorkRequest).toHaveBeenCalledWith(
      experimentId,
      nonGeneExpressionWorkRequest,
      timeout,
      NON_GENE_EXPRESSION_ETAG,
      expect.anything(),
    );
    expect(mockCacheGet).toHaveBeenCalledTimes(1);
    expect(mockCacheSet).toHaveBeenCalledTimes(1);
    expect(mockCacheSet).toHaveBeenCalledWith(NON_GENE_EXPRESSION_ETAG, mockGenesListData);
    expect(mockSeekFromS3).toHaveBeenCalledTimes(2);
    expect(res).toEqual(mockGenesListData);
  });

  it('Throws an error if the dispatched work request throws an error', async () => {
    mockDispatchWorkRequest.mockImplementationOnce(() => { throw new Error('Worker timeout'); });

    await expect(fetchWork(
      experimentId,
      nonGeneExpressionWorkRequest,
      mockReduxState(experimentId),
      { timeout: 10 },
    )).rejects.toThrow();

    expect(mockCacheGet).toHaveBeenCalledTimes(1);
    expect(mockCacheSet).not.toHaveBeenCalled();

    // Only called once when checking for the work result in S3
    expect(mockSeekFromS3).toHaveBeenCalledTimes(1);
  });

  it('does not change ETag if caching is enabled', async () => {
    Storage.prototype.getItem = jest.fn((key) => (key === 'disableCache' ? 'false' : null));

    await fetchWork(
      experimentId,
      nonGeneExpressionWorkRequest,
      mockReduxState(experimentId),
      { timeout },
    );

    expect(mockDispatchWorkRequest).toHaveBeenCalledWith(
      expect.anything(),
      expect.anything(),
      expect.anything(),
      NON_GENE_EXPRESSION_ETAG,
      expect.anything(),
    );
  });

  it('changes ETag if caching is disabled', async () => {
    Storage.prototype.getItem = jest.fn((key) => (key === 'disableCache' ? 'true' : null));

    await fetchWork(
      experimentId,
      nonGeneExpressionWorkRequest,
      mockReduxState(experimentId),
      { timeout },
    );

    expect(mockDispatchWorkRequest).not.toHaveBeenCalledWith(
      expect.anything(),
      expect.anything(),
      expect.anything(),
      NON_GENE_EXPRESSION_ETAG,
      expect.anything(),
    );
  });

  it('Caching is disabled by default if environment is dev', async () => {
    await fetchWork(
      experimentId,
      nonGeneExpressionWorkRequest,
      mockReduxState(experimentId, Environment.DEVELOPMENT),
      { timeout },
    );

    expect(mockDispatchWorkRequest).not.toHaveBeenCalledWith(
      expect.anything(),
      expect.anything(),
      expect.anything(),
      NON_GENE_EXPRESSION_ETAG,
      expect.anything(),
    );
  });

  it('Setting cache to false in development enables cache', async () => {
    Storage.prototype.getItem = jest.fn((key) => (key === 'disableCache' ? 'false' : null));

    await fetchWork(
      experimentId,
      nonGeneExpressionWorkRequest,
      mockReduxState(experimentId, Environment.DEVELOPMENT),
      { timeout },
    );

    expect(mockDispatchWorkRequest).toHaveBeenCalledWith(
      expect.anything(),
      expect.anything(),
      expect.anything(),
      NON_GENE_EXPRESSION_ETAG,
      expect.anything(),
    );
  });
});
