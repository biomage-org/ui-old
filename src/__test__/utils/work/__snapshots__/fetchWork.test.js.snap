// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchWork runs correctly for gene expression work request 1`] = `
{
  "orderedGeneNames": [
    "A",
    "B",
    "C",
    "D",
  ],
  "rawExpression": {
    "datatype": undefined,
    "index": [
      0,
      0,
      2,
      0,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": [
      0,
      1,
      3,
      5,
    ],
    "size": [
      3,
      3,
    ],
    "values": [
      1,
      2,
      5,
      1,
      6,
    ],
  },
  "stats": {
    "rawMean": [
      0.3,
      2.3,
      1.3,
      14.2,
    ],
    "rawStdev": [
      0.4,
      2,
      1,
      1.2,
    ],
    "truncatedMax": [
      1,
      3,
      5,
      0.5,
    ],
    "truncatedMin": [
      0,
      2,
      1,
      0.1,
    ],
  },
  "truncatedExpression": {
    "datatype": undefined,
    "index": [
      0,
      0,
      2,
      0,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": [
      0,
      1,
      3,
      5,
    ],
    "size": [
      3,
      3,
    ],
    "values": [
      1,
      2,
      3,
      0.5,
      1,
    ],
  },
  "zScore": {
    "datatype": undefined,
    "index": [
      0,
      0,
      2,
      0,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": [
      0,
      1,
      3,
      5,
    ],
    "size": [
      3,
      3,
    ],
    "values": [
      1,
      1,
      3,
      3,
      1,
    ],
  },
}
`;
